Add support for non-deterministic games to AlphaZero

egeozsoy opened the pull request:
Addresses the issue #774.

lanctot commented:
@christianjans can you take a look at the LibTorch AZ changes?

christianjans reviewed (commented):
Hi all, sorry for the late response.

Thanks @egeozsoy! This looks great! I'll confess I am not really familiar with AlphaZero applied to non-deterministic games, but I can try and input on some of the changes.

I appreciate you adding an MLP model and adding the `nn_model` variable to the `ModelConfig`. It's now much easier to add new model types. One inevitable repercussion of this is that old models cannot be loaded anymore due to the way we save and load the config as we now save/load an extra parameter (`nn_model`). @lanctot, if there are other people using Libtorch AZ, maybe this would be something to mention in future release notes?

When I implemented this off of the C++ TensorFlow version, I tried to follow Google's C++ style guide (https://google.github.io/styleguide/cppguide.html) as much as possible. This includes things like indentation is 2 spaces, `private`/`public` in class definitions are indented by 1 space, max line length is 80, etc. If possible, this would be a nice-to-have in these changes. But if this is not required by OpenSpiel, then I'm okay with how it is. My suggestions below are just a couple things I found that may help improve readability.

There are two examples in OpenSpiel's C++ [examples](https://github.com/deepmind/open_spiel/tree/master/open_spiel/examples) that relate to Libtorch AZ, namely, [alpha_zero_torch_example.cc](https://github.com/deepmind/open_spiel/blob/master/open_spiel/examples/alpha_zero_torch_example.cc) and [alpha_zero_torch_game_example.cc](https://github.com/deepmind/open_spiel/blob/master/open_spiel/examples/alpha_zero_torch_game_example.cc). If we want these changes to be accessible in these examples, I have outlined some changes that would have to be made:

1. Add "mlp" option: https://github.com/deepmind/open_spiel/blob/a2f209d0921d29fe31957689a217b208897bb705/open_spiel/examples/alpha_zero_torch_example.cc#L29-L30
2. Remove: https://github.com/deepmind/open_spiel/blob/a2f209d0921d29fe31957689a217b208897bb705/open_spiel/examples/alpha_zero_torch_game_example.cc#L179-L180

These changes to these files may make it easier for others to use Libtorch AZ in non-deterministic games and also use the MLP model option. Have you been able to test these changes on non-deterministic games? If so, I would be curious to see how it did.

Let me know if you have any questions or if I can clarify anything.

lanctot commented:
Thank @christianjans !

I will contact the prople I know using it but generally it is ok to be backward-incompatible on the master branch, and we make this clear. For more stability the best is to use a release branch or fork.

Speaking of release we are hoping to release 1.1 in early March which this should make it in for. So then these new changes will be associated with that release branch.

I agree though that the code should adhere to the Google style guide just like the rest of the code base, so please fix these if there are any inconsistencies. I will go through the code tomorrow if you want to give a few examples. Also the fixes to the examples should also be made. Thanks.

egeozsoy commented:
@christianjans Thanks for the review! I now made the modifications you suggested, and also used the google style in my IDE to reformat my changes. I think they now look good. I corrected the argument description to reflect that MLP is also available as an option. 

Also I modified [alpha_zero_torch_game_example.cc](https://github.com/deepmind/open_spiel/blob/master/open_spiel/examples/alpha_zero_torch_game_example.cc) as you suggested. Though it is important to note that, that was not enough to make it work with non deterministic games. Basically in non-deterministic games, actions can be performed by bots or chosen by state outcomes (in chance nodes). This made it hard to work with the current structure, so I modified it a bit. I deciding to commit this version as the previous version just does not work and it does not make sense to have it in the repository. I am open to different ideas to make it work though, if my changes are suboptimal here.

Finally, I successfully trained it to play backgammon, and it is able to beat a random player 100% of the time and vanilla MCTS more than 80% of the time. Interestingly I discovered that using GPU is actually slower when using such a small network (MLP, hidden size 128, num layers 2) so I am now running it on the CPU. Also in my case, the C++ version is only 2 times faster than the python version. Right now I started a long run with the python version, with 500 max simulations, which will take some time. I am looking forward to playing against it to test it :)

elkhrt commented on open_spiel/algorithms/mcts.cc:
```diff
@@ -200,11 +200,12 @@ std::vector<double> dirichlet_noise(int count, double alpha,
 }
 
-MCTSBot::MCTSBot(const Game& game, std::shared_ptr<Evaluator> evaluator,
+MCTSBot::MCTSBot(const Game &game, std::shared_ptr<Evaluator> evaluator,
```
Revert whitespace change (these sorts of changes just add noise to the diff)

egeozsoy replied:
Changed

elkhrt commented on open_spiel/algorithms/mcts.cc:
```diff
@@ -217,7 +218,8 @@ MCTSBot::MCTSBot(const Game& game, std::shared_ptr<Evaluator> evaluator,
       child_selection_policy_(child_selection_policy),
-      evaluator_(evaluator) {
+      evaluator_(evaluator),
+      dont_return_chance_node_{dont_return_chance_node} {
```
Use of {...} is inconsistent with surround lines. Consistency is preferable.

egeozsoy replied:
Use (...) instead

elkhrt commented on open_spiel/algorithms/mcts.cc:
```diff
@@ -271,7 +273,8 @@ std::unique_ptr<State> MCTSBot::ApplyTreePolicy(
   SearchNode* current_node = root;
-  while (!working_state->IsTerminal() && current_node->explore_count > 0) {
+  while ((!working_state->IsTerminal() && current_node->explore_count > 0) ||
+  (working_state->IsChanceNode()) and dont_return_chance_node_) {
```
use && instead of `and` for consistency

egeozsoy replied:
Converted to &&

gdsmithwmi commented:
Hi all.  I'm interested in adapting the non-deterministic backgammon game to an incomplete information game.  I've managed to get the build_and_run_tests.sh to complete successfully under a new installation of Ubuntu.  I am struggling with running backgammon in both python and c++.  Can you share example scripts you ran to train/play backgammon? I would greatly appreciate any help and apologize for my ignorance.

lanctot commented:
Hi @gdsmithwmi,

Are you trying AlphaZero? If so, have you read these docs? https://github.com/deepmind/open_spiel/blob/master/docs/alpha_zero.md

It might be best to either open a new issue or try following up on this thread: https://github.com/deepmind/open_spiel/issues/774  (rather than this closed PR thread).

gdsmithwmi commented:
Thanks @lanctot I should have been clearer, I am trying AlphaZero.
I have read the docs, but I'll poke through again, if that doesn't work I'll try #774.
Thanks for your patience.

lanctot commented:
Of course. Maybe you can clarify about which version of AlphaZero you'd like to run: the python TF version, or the C++-based LibTorch version?

gdsmithwmi commented:
Hi again.  I was able to get the C++ LibTorch version running for
backgammon with the help of the documentation you pointed me to.
The secret is to use the --nn_model='mlp' flag, which I figured out from
reading through the thread.

I would really like to use python/TF as I am more familiar with both.
However, the "generic" alpha_zero.py script does not appear to be able to
handle
games with chance nodes.  I'm almost certainly missing something.

On Fri, Sep 23, 2022 at 1:59 PM lanctot ***@***.***> wrote:

> Of course. Maybe you can clarify about which version of AlphaZero you'd
> like to run: the python TF version, or the C++-based LibTorch version?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/deepmind/open_spiel/pull/776#issuecomment-1256620470>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANHIBF2K4UWHGRYW6WDT373V7YD2VANCNFSM5M5IULCA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lanctot commented:
Hmm, @egeozsoy added support to chance nodes to both C++ and Python MCTS, so I surprised to hear that. Can you post the error?

gdsmithwmi commented:
Sure, thanks for the help.  Here is what I try to run.  And below the
output.  I may also not have cuda installed correctly
as my running C++/torch version also does not appear to be using GPU

export PYTHONPATH=$PYTHONPATH:/home/me/open_spiel
export PYTHONPATH=$PYTHONPATH:/home/me/open_spiel/build/python
echo $PYTHONPATH
python3 open_spiel/python/examples/alpha_zero.py --game backgammon
--nn_model mlp actors 4

2022-09-24 13:55:39.871345: I
tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary
is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the
following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate
compiler flags.
2022-09-24 13:55:39.945596: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0:
cannot open shared object file: No such file or directory
2022-09-24 13:55:39.945613: I
tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart
dlerror if you do not have a GPU set up on your machine.
2022-09-24 13:55:39.960016: E
tensorflow/stream_executor/cuda/cuda_blas.cc:2981] Unable to register
cuBLAS factory: Attempting to register factory for plugin cuBLAS when one
has already been registered
2022-09-24 13:55:40.277131: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot
open shared object file: No such file or directory
2022-09-24 13:55:40.277165: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libnvinfer_plugin.so.7'; dlerror:
libnvinfer_plugin.so.7: cannot open shared object file: No such file or
directory
2022-09-24 13:55:40.277171: W
tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning:
Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU
with TensorRT, please make sure the missing libraries mentioned above are
installed properly.
2022-09-24 13:55:40.946311: I
tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:980] successful NUMA
node read from SysFS had negative value (-1), but there must be at least
one NUMA node, so returning NUMA node zero
2022-09-24 13:55:40.946460: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0:
cannot open shared object file: No such file or directory
2022-09-24 13:55:40.946488: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcublas.so.11'; dlerror: libcublas.so.11: cannot
open shared object file: No such file or directory
2022-09-24 13:55:40.946513: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcublasLt.so.11'; dlerror: libcublasLt.so.11:
cannot open shared object file: No such file or directory
2022-09-24 13:55:40.946536: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcufft.so.10'; dlerror: libcufft.so.10: cannot open
shared object file: No such file or directory
2022-09-24 13:55:40.946560: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcurand.so.10'; dlerror: libcurand.so.10: cannot
open shared object file: No such file or directory
2022-09-24 13:55:40.946582: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcusolver.so.11'; dlerror: libcusolver.so.11:
cannot open shared object file: No such file or directory
2022-09-24 13:55:40.946603: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcusparse.so.11'; dlerror: libcusparse.so.11:
cannot open shared object file: No such file or directory
2022-09-24 13:55:40.946625: W
tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not
load dynamic library 'libcudnn.so.8'; dlerror: libcudnn.so.8: cannot open
shared object file: No such file or directory
2022-09-24 13:55:40.946631: W
tensorflow/core/common_runtime/gpu/gpu_device.cc:1934] Cannot dlopen some
GPU libraries. Please make sure the missing libraries mentioned above are
installed properly if you would like to use GPU. Follow the guide at
https://www.tensorflow.org/install/gpu for how to download and setup the
required libraries for your platform.
Skipping registering GPU devices...
Starting game backgammon
Traceback (most recent call last):
  File "/home/grant/open_spiel/open_spiel/python/examples/alpha_zero.py",
line 99, in <module>
    app.run(main)
  File "/home/grant/.local/lib/python3.10/site-packages/absl/app.py", line
308, in run
    _run_main(main, args)
  File "/home/grant/.local/lib/python3.10/site-packages/absl/app.py", line
254, in _run_main
    sys.exit(main(argv))
  File "/home/grant/open_spiel/open_spiel/python/examples/alpha_zero.py",
line 94, in main
    alpha_zero.alpha_zero(config)
  File
"/home/grant/open_spiel/open_spiel/python/algorithms/alpha_zero/alpha_zero.py",
line 514, in alpha_zero
    raise ValueError("Game must be deterministic.")
ValueError: Game must be deterministic.

On Fri, Sep 23, 2022 at 3:20 PM lanctot ***@***.***> wrote:

> Hmm, @egeozsoy <https://github.com/egeozsoy> added support to chance
> nodes to both C++ and Python MCTS, so I surprised to hear that. Can you
> post the error?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/deepmind/open_spiel/pull/776#issuecomment-1256688653>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANHIBF22EAXUCDL64XF7KS3V7YNJNANCNFSM5M5IULCA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lanctot commented:
Ah. At this point I think it's a relatively simple change because the Python MCTS now properly supports chance nodes.

I think @egeozsoy added the actual support for chance also in the self-play, because I see it right here: https://github.com/deepmind/open_spiel/blob/3dc0ca65c90ad3d3642a4435781722744a1356b9/open_spiel/python/algorithms/alpha_zero/alpha_zero.py#L212

So I think it should be as simple as removing that check (on lines 513 and 514). But if those lines are still there, it probably means the case of chance games was never tested. If it runs for you, please let us know.. and maybe submit a Pull Request that removes those two lines.
